packages/flutter/lib/src/foundation/assertions.dart

```diff
@@ -147,41 +285,98 @@ class FlutterErrorDetails {
+          message = message.substring(prefix.length);
+        properties.add(ErrorDescription('$message'));
+      }
+    }
+
+    final Iterable<String> stackLines = (stack != null) ? stack.toString().trimRight().split('\n') : null;
+    if ((exception is AssertionError) && (exception is! FlutterError)) {
+      bool ourFault = true;
+      if (stackLines != null) {
+        final List<String> stackList = stackLines.take(2).toList();
+        if (stackList.length >= 2) {
+          // TODO(ianh): This has bitrotted and is no longer matching. https://github.com/flutter/flutter/issues/4021
+          final RegExp throwPattern = RegExp(
+              r'^#0 +_AssertionError._throwNew \(dart:.+\)$');
+          final RegExp assertPattern = RegExp(
+              r'^#1 +[^(]+ \((.+?):([0-9]+)(?::[0-9]+)?\)$');
+          if (throwPattern.hasMatch(stackList[0])) {
+            final Match assertMatch = assertPattern.firstMatch(stackList[1]);
+            if (assertMatch != null) {
+              assert(assertMatch.groupCount == 2);
+              final RegExp ourLibraryPattern = RegExp(r'^package:flutter/');
+              ourFault = ourLibraryPattern.hasMatch(assertMatch.group(1));
+            }
+          }
+        }
+      }
+      if (ourFault) {
+        properties.add(DiagnosticsNode.message(''));
+        properties.add(ErrorHint(
+          'Either the assertion indicates an error in the framework itself, or we should '
+          'provide substantially more information in this error message to help you determine '
+          'and fix the underlying cause.\n'
+          'In either case, please report this assertion by filing a bug on GitHub:\n'
+          '  https://github.com/flutter/flutter/issues/new?template=BUG.md'
+        ));
       }
-      buffer.writeAll(stackLines, '\n');
     }
-    return buffer.toString().trimRight();
+    if (stack != null) {
+      properties.add(DiagnosticsNode.message(''));
+      properties.add(DiagnosticsStackTrace('When the exception was thrown, this was the stack', stack, stackFilter: stackFilter));
+      if (informationCollector != null) //ignore: deprecated_member_use_from_same_package
+        properties.add(DiagnosticsNode.message(''));
+    }
+    if (informationCollector != null) {
+      final List<DiagnosticsNode> information = <DiagnosticsNode>[];
+      informationCollector(information);
+      information.forEach(properties.add);
+    }
+  }
+
+  @override
+  String toStringShort() {
+    return library != null ? 'Exception Caught By $library' : 'Exception Caught';
+  }
+
+  @override
+  String toString({DiagnosticLevel minLevel = DiagnosticLevel.debug}) {
+    return toDiagnosticsNode(style: DiagnosticsTreeStyle.error).toStringDeep(minLevel: minLevel);
   }
 }
 
 /// Error class used to report Flutter-specific assertion failures and
 /// contract violations.
-class FlutterError extends AssertionError {
+class FlutterError extends AssertionError with DiagnosticableTreeMixin {
```

Comment:
we should make this just `implements AssertionError` rather than `extends` since we don't actually use any of the features of `AssertionError` any more.

Reply:
done